src/goto-instrument/contracts/utils.cpp

```diff
@@ -210,3 +214,71 @@ bool is_auxiliary_decl_dead_or_assign(
   return false;
 }
 
+void simplify_gotos(goto_programt &goto_program, namespacet &ns)
+{
+  for(auto &instruction : goto_program.instructions)
+  {
+    if(instruction.is_goto())
+    {
+      const auto &condition = instruction.get_condition();
+      const auto &simplified = simplify_expr(condition, ns);
+      if(simplified.is_false())
+        instruction.turn_into_skip();
+    }
+  }
+}
+
+bool is_loop_free(
+  const goto_programt &goto_program,
+  namespacet &ns,
+  messaget &log)
+{
+  // create cfg from instruction list
+  cfg_baset<empty_cfg_nodet> cfg;
+  cfg(goto_program);
+
```

Comment:
You wouldn't have quite as elaborate error messages, but wouldn't a simple test of all instructions for `is_backwards_goto()` do the trick?